compile/ivy/src/main/scala/sbt/compiler/ComponentCompiler.scala

```diff
@@ -64,4 +94,129 @@ class ComponentCompiler(compiler: RawCompiler, manager: ComponentManager) {
+ * the compiler interface sources using Ivy.
+ * The compiled classes are cached using the provided component manager according
+ * to the actualVersion field of the RawCompiler.
+ */
+private[compiler] class IvyComponentCompiler(compiler: RawCompiler, manager: ComponentManager, ivyConfiguration: IvyConfiguration, log: Logger) {
+  import ComponentCompiler._
+
+  private val ivySbt: IvySbt = new IvySbt(ivyConfiguration)
+  private val sbtVersion = ComponentManager.version
+
+  def apply(id: String): File = {
+    val binID = binaryID(id)
+    manager.file(binID)(new IfMissing.Define(true, compileAndInstall(id, binID)))
+  }
+
+  private def binaryID(id: String): String = {
+    val base = id + binSeparator + compiler.scalaInstance.actualVersion
+    base + "__" + javaVersion
+  }
+
+  private def compileAndInstall(id: String, binID: String): Unit = {
+    def interfaceSources(moduleVersions: Seq[String]): Iterable[File] =
+      moduleVersions match {
+        case Seq() =>
+          def getAndDefineDefaultSources() =
+            update(getModule(id))(_.getName endsWith "-sources.jar") map { sourcesJar =>
+              manager.define(id, sourcesJar)
+              sourcesJar
+            } getOrElse (throw new InvalidComponent(s"Couldn't retrieve default sources: module '$id'"))
+
+          log.debug(s"Fetching default sources: module '$id'")
+          manager.files(id)(new IfMissing.Fallback(getAndDefineDefaultSources()))
+
+        case version +: rest =>
+          val moduleName = s"${id}_$version"
+          def getAndDefineVersionSpecificSources() =
+            update(getModule(moduleName))(_.getName endsWith "-sources.jar") map { sourcesJar =>
+              manager.define(moduleName, sourcesJar)
+              sourcesJar
+            } getOrElse interfaceSources(rest)
+
+          log.debug(s"Fetching version-specific sources: module '$moduleName'")
+          manager.files(moduleName)(new IfMissing.Fallback(getAndDefineVersionSpecificSources()))
+      }
+    IO.withTemporaryDirectory { binaryDirectory =>
+
+      val targetJar = new File(binaryDirectory, s"$binID.jar")
+      val xsbtiJars = manager.files(xsbtiID)(IfMissing.Fail)
+
+      val sourceModuleVersions = cascadingSourceModuleVersions(compiler.scalaInstance.actualVersion)
+      AnalyzingCompiler.compileSources(interfaceSources(sourceModuleVersions), targetJar, xsbtiJars, id, compiler, log)
+
+      manager.define(binID, Seq(targetJar))
+
+    }
+  }
+
+  /**
+   * Returns a dummy module that depends on "org.scala-sbt" % `id` % `sbtVersion`.
+   * Note: Sbt's implementation of Ivy requires us to do this, because only the dependencies
+   *       of the specified module will be downloaded.
+   */
+  private def getModule(id: String): ivySbt.Module = {
+    val dummyID = ModuleID(xsbti.ArtifactInfo.SbtOrganization, scala.util.Random.alphanumeric take 20 mkString "", sbtVersion, Some("component"))
```

Comment:
I'd suggest against using the actual "org.scala-sbt", but use "org.scala-sbt.temp" instead. That's what I do for cached resolution.